### PR TITLE
[BEAM-10101] Add HttpIO / HttpFileSystem (Python)

### PR DESCRIPTION
**Pull request by epicfaace**
Add HttpIO (and a related HttpFileSystem and HttpsFileSystem), which can download files from a particular http:// or https:// URL. HttpIO cannot upload / write to files, though, because there's no standardized way to write to files using HTTP.

Sample usage:

```python
        (
            p
            | ReadFromText("https://raw.githubusercontent.com/apache/beam/5ff5313f0913ec81d31ad306400ad30c0a928b34/NOTICE")
            | WriteToText("output.txt", shard_name_template="", num_shards=0)
        )
```

R: @pabloem

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
@aaltay can you please review this one or pass to someone who can. thx!

**Comment by pabloem** [comment]
I'll be happy to take a look at this

**Comment by pabloem** [comment]
ok just looking at this now...

**Comment by pabloem** [comment]
retest this please

**Comment by pabloem** [comment]
@epicfaace lmk what are your plans for this PR

**Comment by epicfaace** [comment]
I'll be making changes soon.

**Comment by epicfaace** [comment]
@pabloem I've addressed your changes and also made the implementation a bit cleaner; please take a look.

**Comment by pabloem** [comment]
retest this please

**Comment by pabloem** [comment]
retest this please

**Comment by pabloem** [comment]
Run Portable_Python PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Run PythonFormatter PreCommit

**Comment by pabloem** [comment]
Run PythonLint PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by epicfaace** [comment]
@pabloem anything you need from me on this one?

**Comment by pabloem** [comment]
sorry about the delay - it looks like many of the tests are failing: https://ci-beam.apache.org/job/beam_PreCommit_Python_Phrase/1920/
Do you htink you could fix those so we can review?

File: sdks/python/apache_beam/io/httpio.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+from apache_beam.io.filesystemio import DownloaderStream
+from apache_beam.internal.http_client import get_new_http
+import sys
+from httplib2 import HttpLib2Error
+
+REQUEST_FAILED_ERROR_MSG = "HTTP request failed for URL {}: {}"
+UNEXPECTED_STATUS_CODE_ERROR_MSG = "Unexpected status code received for URL {}: {} {}"
+
+
+class HttpIO(object):
```
Comment (epicfaace) [inline sdks/python/apache_beam/io/httpio.py:46]:
```suggestion
class HttpIO:
```

**Comment by epicfaace** [comment]
I'm not sure why exactly those failures are happening, since `HttpIO` _does_ take `client` as a constructor.

**Comment by pabloem** [comment]
okay let me figure out why that's happening...

**Comment by epicfaace** [comment]
Oh, it appears the test failures are only on Python 2 -- perhaps I didn't define the class properly for Python 2

**Comment by pabloem** [comment]
should I look into it? Will you? @epicfaace LMK : )

**Comment by epicfaace** [comment]
I'll take a look!

**Comment by epicfaace** [comment]
Still need to take a look -- not stale

**Comment by epicfaace** [comment]
Todo

**Comment by TheNeuralBit** [comment]
If the test failures are only Python 2 maybe this will just work now that we've dropped Python 2 support :)

**Pull request closed without merging**